[ResponseOps] Update flapping logic order to determine whether an alert is flapping after it's returned for notification

doakalexi opened the pull request:
Resolves https://github.com/elastic/kibana/issues/151135

## Summary

Moved the logic around to check flapping after an alert is already returned for notification. I also updated the event log test cases to match examples in the rfc.


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### To Verify
The new test cases capture this example but if you want to test it locally:
- Create a connector and make sure to save the connector id
```
curl -X POST -u {username}:{password} "http://localhost:5601/api/actions/connector" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d'
{
  "connector_type_id": ".server-log",
  "name": "server log",
  "config": {},
  "secrets": {}
}'
```
- Create the rule and let it run
```
curl -X POST -u {username}:{password} "http://localhost:5601/api/alerting/rule/" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d'
{
  "rule_type_id": "example.pattern",
  "name": "pattern",
  "schedule": {
    "interval": "5s"
  },
  "actions": [
    { "group": "default", "id": {connector id}, "params": { "message": "{{alert.id}} active on run {{context.runs}} step {{context.patternIndex}} flapping {{alert.flapping}}"}},
    { "group": "recovered", "id": {connector id}, "params": { "message": "{{alert.id}} recovered on run flapping {{alert.flapping}}"}}
  ],
  "consumer": "alerts",
  "tags": [],
  "notify_when": "onActionGroupChange",
  "params": {
    "patterns": {
      "instA": " a - - a - a - a - a - - - - - - - - "
    }
  }
}'

```
- Verify that you see the following, we want to make sure that aren't missing any notifications
![image](https://user-images.githubusercontent.com/109488926/221615405-48061d10-4b80-4b98-812a-6951da4178da.png)


<!--ONMERGE {"backportTargets":["8.7"]} ONMERGE-->

doakalexi committed:
Updating flapping logic order

doakalexi committed:
Updating tests

doakalexi committed:
Adding a new test

doakalexi committed:
Merge branch 'main' of github.com:elastic/kibana into alerting/update-flapping-logic-order

doakalexi committed:
Merge branch 'main' of github.com:elastic/kibana into alerting/update-flapping-logic-order

doakalexi committed:
Removing only

doakalexi committed:
Fixing test failure

doakalexi committed:
Call flapping in the task runner

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

doakalexi committed:
Merge branch 'main' into alerting/update-flapping-logic-order

doakalexi committed:
Merge branch 'main' into alerting/update-flapping-logic-order

maryam-saeidi reviewed (commented):
> Moved the logic around to check flapping after an alert is already returned for notification.

I am trying to understand this description. How was the flapping status generation different in the previous implementation?

doakalexi commented:
> > Moved the logic around to check flapping after an alert is already returned for notification.
> 
> I am trying to understand this description. How was the flapping status generation different in the previous implementation?

The flapping status generation didn't change. I just moved things around in this pr to match what the RFC is expecting. Before we would check to see whether an alert was flapping and then determine if that alert is going to be reported as a notification to users. The flapping status affects whether an alert is reported to the user, and so by checking flapping first we  were missing notifications in examples from the RFC that were expecting notifications. In this pr, I moved the flapping check to be after we determine which alert is going to be reported to the user which aligns with the RFC

maryam-saeidi reviewed (approved):
AO changes LGTM

pmuellr reviewed (approved):
LGTM, testing it locally seems to work as expected
